### DISCUSSION TITLE
Exported YOLO model has fewer classes than defined in dataset.yaml

### DISCUSSION THREAD

**Discussion by mandelacruz** [ultralytics/ultralytics#22096] [Q&A] [answered]
Hi, I’m running into an issue with my model export.

- I annotated my dataset using Roboflow.
- I trained, validated, and tested the model in Google Colab, and also tried running it in Ultralytics Hub.
- I trained multiple times from scratch using yolo11n.yaml and also fine-tuned a pretrained model with yolo11n.pt.
- During training, everything works fine — the model detects images accurately, and the number of classes matches my dataset.yaml (7 classes).

Problem:
- When I export the model (to .pt or .tflite), it only shows 6 classes instead of 7.
- The output tensor shape becomes [1, 8400, 11] instead of the expected [1, 8400, 12].
- This also carries over when I integrate the exported .tflite model into my app project.

Has anyone encountered this before? Could this be an export bug, or am I missing a step when defining the classes/dataset?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mandelacruz, thanks for the detailed report and for using Ultralytics 🚀 This is an automated response — an Ultralytics engineer will follow up with you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report (it sounds like an export-related issue), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. To speed things up, please include:
-Exact `dataset.yaml` (with names list length)
-The exact train command(s) you ran (e.g., using YOLO11 configs like `yolo11n.yaml` or weights like `yolo11n.pt`)
-The exact export command(s) and args you used for `.pt` and `.tflite`
-Environment details: `ultralytics` version, OS, Python, PyTorch, and where you trained/exported (Colab, local, Ultralytics HUB)
-Key logs from train/val/export and a tiny sample of data if possible
-A short code snippet showing how you’re loading and checking model outputs and shapes
-Expected vs. observed output shapes/classes

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by mandelacruz** [comment]
Dataset Configuration (datasets.yaml)

```
nc: 7
names: ['ASV 1', 'ASV 2', 'ASV 3', 'ASV 4', 'ASV 5', 'ASV 6', 'ASV 7']
```

Training (Google Colab)

```
!pip install ultralytics roboflow

from ultralytics import YOLO
from roboflow import Roboflow

# Roboflow dataset
rf = Roboflow(api_key="API_KEY")
project = rf.workspace("WORKSPACE").project("PROJECT")
version = project.version(3)
dataset = version.download("yolov11")

# Train
!yolo task=detect mode=train data={dataset.location}/data.yaml model=yolo11n.yaml epochs=100 imgsz=640

# Validate
!yolo task=detect mode=val model=/PATH/TO/best.pt data={dataset.location}/data.yaml

# Predict
!yolo task=detect mode=predict model=/PATH/TO/best.pt source={dataset.location}/train/images save=True
```


Export (Google Colab)

```
from ultralytics import YOLO

model = YOLO("/PATH/TO/best.pt")
model.export(format="tflite")

# Test exported TFLite model
!yolo mode=predict task=detect model=/PATH/TO/best_float32.tflite imgsz=640 source={dataset.location}/train/images save=True
!yolo mode=val task=detect model=/PATH/TO/best_float32.tflite imgsz=640 data=/content/ASV-3/data.yaml
```


Expected output shapes: [1,8400,12]
Observed output shapes: [1,11,8400]

Expected classes: 7
Observed classes: 6

Sample Predictions
![val_batch1_pred](https://github.com/user-attachments/assets/44a85166-e0e2-4dea-8f72-cd331abbf061)
![val_batch0_pred (1)](https://github.com/user-attachments/assets/793977c3-9790-47e8-9ef8-370a9fb871e3)

**Comment by Y-T-G** [comment, marked as answer]
> The output tensor shape becomes [1, 8400, 11] instead of the expected [1, 8400, 12].

It's correct. You have incorrect assumption about what the output values are.

https://www.reddit.com/r/Ultralytics/comments/1jwl6se/ultralytics_postprocessing_guide/
